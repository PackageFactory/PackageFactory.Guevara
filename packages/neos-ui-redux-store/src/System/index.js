import {createAction} from 'redux-actions';

import {handleActions} from '@neos-project/utils-redux';

const BOOT = '@neos/neos-ui/System/BOOT';
const INIT = '@neos/neos-ui/System/INIT';
const READY = '@neos/neos-ui/System/READY';

//
// Export the action types
//
export const actionTypes = {
    BOOT,
    INIT,
    READY
};

const boot = createAction(BOOT);
const init = createAction(INIT, state => state);
const ready = createAction(READY);

//
// Export the actions
//
export const actions = {
    boot,
    init,
    ready
};

//
// Export the reducer
//
export const reducer = handleActions({
});

//
// Export the selectors
//
export const selectors = {};
