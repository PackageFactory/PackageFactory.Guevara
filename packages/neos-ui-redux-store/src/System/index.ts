import produce from 'immer';
import {$get} from 'plow-js';
import {action as createAction, ActionType as ActionTypeHelper} from 'typesafe-actions';

const BOOT = '@neos/neos-ui/System/BOOT';
const INIT = '@neos/neos-ui/System/INIT';
const READY = '@neos/neos-ui/System/READY';
const AUTHENTICATION_TIMEOUT = '@neos/neos-ui/System/AUTHENTICATION_TIMEOUT';
const REAUTHENTICATION_SUCCEEDED = '@neos/neos-ui/System/REAUTHENTICATION_SUCCEEDED';

//
// Export the action types
//
export const actionTypes = {
    AUTHENTICATION_TIMEOUT,
    BOOT,
    INIT,
    READY,
    REAUTHENTICATION_SUCCEEDED
};

interface State {
    readonly authenticationTimeout: boolean;
}

const boot = () => createAction(BOOT);
const init = (state: State) => createAction(INIT, state);
const ready = () => createAction(READY);
const authenticationTimeout = () => createAction(AUTHENTICATION_TIMEOUT);
const reauthenticationSucceeded = () => createAction(REAUTHENTICATION_SUCCEEDED);

//
// Export the actions
//
export const actions = {
    authenticationTimeout,
    boot,
    init,
    ready,
    reauthenticationSucceeded
};

export type ActionType = ActionTypeHelper<typeof actions>;

const defaultState: State = {
    authenticationTimeout: false
};

//
// Export the reducer
//
export const reducer = (state: State = defaultState, action: ActionType) => {
    return produce(state, draft => {
        switch (action.type) {
            case INIT:
                draft.authenticationTimeout = false;
                break;
            case AUTHENTICATION_TIMEOUT:
                draft.authenticationTimeout = true;
                break;
            case REAUTHENTICATION_SUCCEEDED:
                draft.authenticationTimeout = false;
                break;
        }
    });
};

//
// Export the selectors
//
export const selectors = {
    authenticationTimeout: (state: any) => $get(['system'], state).authenticationTimeout
};
