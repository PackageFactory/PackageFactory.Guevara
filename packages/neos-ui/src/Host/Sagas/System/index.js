import {take, put} from 'redux-saga/effects';
import {delay, discover} from 'Shared/Utilities/Promises';
import {actionTypes, actions} from 'Host/Redux/index';
import initializeJSAPI from 'API/index';

let injectStore = null;
export const getStore = discover(function * () {
    const store = yield new Promise(resolve => {
        injectStore = resolve;
    });

    return store;
});

export const getAppContainer = discover(function * () {
    const appContainer = yield new Promise(resolve => {
        document.addEventListener('DOMContentLoaded', () => {
            resolve(document.getElementById('appContainer'));
        });
    });

    return appContainer;
});

export const getCsrfToken = discover(function * () {
    const appContainer = yield getAppContainer;

    return appContainer.dataset.csrfToken;
});

export const getSystemEnv = discover(function * () {
    const appContainer = yield getAppContainer;

    return appContainer.dataset.env;
});

export const getServerState = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="initialState"]').innerHTML);
});

export const getConfiguration = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="configuration"]').innerHTML);
});

export const getMenu = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="menu"]').innerHTML);
});

export const getTranslations = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="translations"]').innerHTML);
});

export const getNeos = discover(function * () {
    const csrfToken = yield getCsrfToken;
    const systemEnv = yield getSystemEnv;

    const neos = initializeJSAPI(window, {
        csrfToken,
        systemEnv
    });

    return neos;
});

function * bootSaga(store) {
    injectStore(store);

    yield take(actionTypes.System.BOOT);

    const serverState = yield getServerState;

    //
    // Kick off initialization for redux, other sagas can
    // listen to this action and start their initialization tasks
    //
    yield put(actions.System.init(serverState));

    //
    // Just make sure that everybody does their initialization
    // homework
    //
    yield delay(0);

    //
    // Inform everybody, that we're ready to rock
    //
    yield put(actions.System.ready());
}

export const sagas = [
    bootSaga
];
