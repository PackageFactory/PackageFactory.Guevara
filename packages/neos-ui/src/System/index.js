import {discover} from '@neos-project/utils-helpers';
import {initializeJsAPI} from '@neos-project/neos-ui-backend-connector';

export const getAppContainer = discover(function * () {
    const appContainer = yield new Promise(resolve => {
        document.addEventListener('DOMContentLoaded', () => {
            resolve(document.getElementById('appContainer'));
        });
    });

    return appContainer;
});

export const getCsrfToken = discover(function * () {
    const appContainer = yield getAppContainer;

    return appContainer.dataset.csrfToken;
});

export const getSystemEnv = discover(function * () {
    const appContainer = yield getAppContainer;

    return appContainer.dataset.env;
});

export const getServerState = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="initialState"]').innerHTML);
});

export const getConfiguration = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="configuration"]').innerHTML);
});

export const getNodeTypes = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="nodeTypes"]').innerHTML);
});

export const getMenu = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="menu"]').innerHTML);
});

export const getTranslations = discover(function * () {
    const appContainer = yield getAppContainer;

    return JSON.parse(appContainer.querySelector('[data-json="translations"]').innerHTML);
});

export const getNeos = discover(function * () {
    const csrfToken = yield getCsrfToken;
    const systemEnv = yield getSystemEnv;

    const neos = initializeJsAPI(window, {
        csrfToken,
        systemEnv
    });

    return neos;
});
