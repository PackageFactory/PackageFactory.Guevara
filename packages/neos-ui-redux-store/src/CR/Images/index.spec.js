import Immutable from 'immutable';
import {$all, $set, $get} from 'plow-js';

import {actionTypes, actions, reducer, selectors} from './index.js';
import {actionTypes as system} from '../../System/index';

test(`should export actionTypes`, () => {
    expect(actionTypes).not.toBe(undefined);
    expect(typeof (actionTypes.START_LOADING)).toBe('string');
    expect(typeof (actionTypes.FINISH_LOADING)).toBe('string');
});

test(`should export action creators`, () => {
    expect(actions).not.toBe(undefined);
    expect(typeof (actions.startLoading)).toBe('function');
    expect(typeof (actions.finishLoading)).toBe('function');
});

test(`should export a reducer`, () => {
    expect(reducer).not.toBe(undefined);
    expect(typeof (reducer)).toBe('function');
});

test(`should export selectors`, () => {
    expect(selectors).not.toBe(undefined);
});

test(`The reducer should create a valid initial state`, () => {
    const state = new Immutable.Map({});
    const nextState = reducer(state, {
        type: system.BOOT
    });

    expect(nextState).toMatchSnapshot();
});

test(`START_LOADING should set status to LOADING for the given image`, () => {
    const state = reducer(new Immutable.Map({}), {
        type: system.BOOT
    });
    const action = actions.startLoading('88525b30-59a6-4f97-9a79-b52172b1815b');
    const nextState = reducer(state, action);

    expect($get('cr.images.byUuid.88525b30-59a6-4f97-9a79-b52172b1815b.status', nextState))
        .toBe('LOADING');
});

test(`FINISH_LOADING should set status to LOADED for the given image`, () => {
    const state = Immutable.fromJS($all(
        $set('cr.images.byUuid.9fd46b81-5f95-4c23-8b5d-871b63b10ac9.status', 'LOADING'),
        {}
    ));
    const action = actions.finishLoading('9fd46b81-5f95-4c23-8b5d-871b63b10ac9', {});
    const nextState = reducer(state, action);

    expect($get('cr.images.byUuid.9fd46b81-5f95-4c23-8b5d-871b63b10ac9.status', nextState))
        .toBe('LOADED');
});

test(`FINISH_LOADING should loaded data for the given image`, () => {
    const state = Immutable.fromJS($all(
        $set('cr.images.byUuid.0db0937d-2ee2-4fce-a78f-17adbfdb2fbe.status', 'LOADING'),
        {}
    ));
    const action = actions.finishLoading('0db0937d-2ee2-4fce-a78f-17adbfdb2fbe', {
        some: 'value',
        another: 'different value'
    });
    const nextState = reducer(state, action);

    expect(nextState).toMatchSnapshot();
});
