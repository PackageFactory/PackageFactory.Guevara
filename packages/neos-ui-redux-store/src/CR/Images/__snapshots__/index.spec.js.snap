// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FINISH_LOADING should loaded data for the given image 1`] = `
Immutable.Map {
  "cr": Immutable.Map {
    "images": Immutable.Map {
      "byUuid": Immutable.Map {
        "0db0937d-2ee2-4fce-a78f-17adbfdb2fbe": Immutable.OrderedMap {
          "some": "value",
          "another": "different value",
          "status": "LOADED",
        },
      },
    },
  },
}
`;

exports[`The reducer should create a valid initial state 1`] = `
Immutable.Map {
  "cr": Immutable.Map {
    "images": Immutable.Map {
      "byUuid": Immutable.Map {},
    },
  },
}
`;
